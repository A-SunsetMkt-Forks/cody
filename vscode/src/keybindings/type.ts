import type { CommentArray } from 'comment-json'

export type KeybindingsContent = CommentArray<Record<string, string>> | null | undefined
