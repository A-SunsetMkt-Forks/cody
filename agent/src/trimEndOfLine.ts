export function trimEndOfLine(text: string | undefined): string {
    if (text === undefined) {
        return ''
    }
    return text
        .trim()
        .split('\n')
        .map(line => line.trimEnd())
        .join('\n')
}
