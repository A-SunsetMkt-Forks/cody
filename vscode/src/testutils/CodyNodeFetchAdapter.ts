import FetchAdapter from '@pollyjs/adapter-fetch'
import type { Request } from '@pollyjs/core'

export class CodyNodeFetchAdapter extends FetchAdapter {
    public async onRequest(request: Request): Promise<void> {
        if (request.body) {
            request.body = request.body
                .replaceAll(/`([^`]*)(cody-vscode-shim-test[^`]*)`/g, '`$2`')
                .replaceAll(/(\\\\)(\w)/g, '/$2')
        }

        return super.onRequest(request)
    }

    public async onRequestFailed(request: Request): Promise<void> {
        return super.onRequestFailed(request)
    }
}
