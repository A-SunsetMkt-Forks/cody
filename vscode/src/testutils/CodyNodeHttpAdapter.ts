import NodeHttpAdapter from '@pollyjs/adapter-node-http'
import type { Request } from '@pollyjs/core'

export class CodyNodeHttpAdapter extends NodeHttpAdapter {
    public async onRequest(request: Request): Promise<void> {
        if (request.body) {
            try {
                request.body = request.body
                    .replaceAll(/`([^`]*)(cody-vscode-shim-test[^`]*)`/g, '`$2`')
                    .replaceAll(/(\\\\)(\w)/g, '/$2')
            } catch (e) {
                console.error('CodyNodeHttpAdapter.onRequest', e)
            }
        }

        return super.onRequest(request)
    }
}
