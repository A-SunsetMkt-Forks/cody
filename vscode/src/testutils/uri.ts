import { URI, Utils } from 'vscode-uri'
import { UriComponents } from 'vscode-uri/lib/umd/uri'

export class Uri {
    public static parse(value: string, strict?: boolean): Uri {
        return new Uri(URI.parse(value, strict))
    }

    public static file(path: string): URI {
        return new Uri(URI.file(path))
    }

    public static joinPath(base: Uri, ...pathSegments: string[]): Uri {
        return new Uri(Utils.joinPath(base.uri, ...pathSegments))
    }

    public static from(components: {
        readonly scheme: string
        readonly authority?: string
        readonly path?: string
        readonly query?: string
        readonly fragment?: string
    }): Uri {
        return new Uri(URI.from(components))
    }

    private uri: URI

    private constructor(componentsOrUri: UriComponents | URI) {
        if (componentsOrUri instanceof URI) {
            this.uri = componentsOrUri
        } else {
            this.uri = URI.from(componentsOrUri)
        }
    }

    get scheme() {
        return this.uri.scheme
    }

    get authority() {
        return this.uri.authority
    }
    get path() {
        return this.uri.path
    }

    get query() {
        return this.uri.query
    }

    get fragment() {
        return this.uri.fragment
    }

    get fsPath() {
        return this.uri.fsPath
    }

    public with(change: {
        scheme?: string
        authority?: string
        path?: string
        query?: string
        fragment?: string
    }): Uri {
        return Uri.from({
            scheme: change.scheme || this.scheme,
            authority: change.authority || this.authority,
            path: change.path || this.path,
            query: change.query || this.query,
            fragment: change.fragment || this.fragment,
        })
    }

    public toString(skipEncoding?: boolean): string {
        return this.uri.toString(skipEncoding)
    }

    public toJSON(): any {
        return {
            scheme: this.scheme,
            authority: this.authority,
            path: this.path,
            query: this.query,
            fragment: this.fragment,
        }
    }
}
