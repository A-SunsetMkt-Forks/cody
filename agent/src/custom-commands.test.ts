import path from 'node:path'
import { afterAll, beforeAll, describe, expect, it } from 'vitest'
import { TESTING_CREDENTIALS } from '../../vscode/src/testutils/testing-credentials'
import { TestClient } from './TestClient'
import { TestWorkspace } from './TestWorkspace'
import { explainPollyError } from './explainPollyError'
import type { CustomChatCommandResult, CustomEditCommandResult } from './protocol-alias'
import { trimEndOfLine } from './trimEndOfLine'

describe('Custom Commands', () => {
    const workspace = new TestWorkspace(path.join(__dirname, '__tests__', 'example-ts'))
    const animalUri = workspace.file('src', 'animal.ts')
    const sumUri = workspace.file('src', 'sum.ts')
    const client = TestClient.create({
        workspaceRootUri: workspace.rootUri,
        name: 'customCommandsClient',
        credentials: TESTING_CREDENTIALS.dotcom,
    })

    beforeAll(async () => {
        await workspace.beforeAll()
        await client.beforeAll()
        await client.request('command/execute', { command: 'cody.search.index-update' })
    })

    afterAll(async () => {
        await workspace.afterAll()
        await client.afterAll()
    })

    // Note: needs to be the first test case so that we can control over
    // what tabs are open here.
    it('commands/custom, chat command, open tabs context', async () => {
        const trickyLogicUri = workspace.file('src', 'trickyLogic.ts')
        await client.openFile(workspace.file('src', 'squirrel.ts'))
        await client.openFile(workspace.file('src', 'sum.ts'))
        await client.openFile(workspace.file('src', 'TestLogger.ts'))
        await client.openFile(trickyLogicUri)

        const result = (await client.request('commands/custom', {
            key: '/countTabs',
        })) as CustomChatCommandResult
        expect(result.type).toBe('chat')
        const lastMessage = await client.firstNonEmptyTranscript(result?.chatResult as string)
        expect(trimEndOfLine(lastMessage.messages.at(-1)?.text ?? '')).toMatchInlineSnapshot(
            `
          "Based on the file paths you provided, the files are:

          1. src/TestLogger.ts
          2. src/sum.ts
          3. src/squirrel.ts
          4. src/trickyLogic.ts"
        `
        )
    }, 30_000)

    // CODY-1766: disabled because the generated output is too unstable
    it.skip('commands/custom, chat command, adds argument', async () => {
        await client.openFile(animalUri)
        const result = (await client.request('commands/custom', {
            key: '/translate Python',
        })) as CustomChatCommandResult
        expect(result.type).toBe('chat')
        const lastMessage = await client.firstNonEmptyTranscript(result?.chatResult as string)
        expect(trimEndOfLine(lastMessage.messages.at(-1)?.text ?? '')).toMatchInlineSnapshot(
            `
              "Here's the Python equivalent of the provided TypeScript interface:

              \`\`\`python
              class Animal:
                  def __init__(self, name: str, is_mammal: bool):
                      self.name = name
                      self.is_mammal = is_mammal

                  def make_animal_sound(self) -> str:
                      raise NotImplementedError("Subclasses must implement make_animal_sound method")
              \`\`\`

              In Python, we don't have the concept of an interface like in TypeScript, so we use a base class with abstract methods. The \`Animal\` class has the following:

              1. An \`__init__\` method that takes in \`name\` (a string) and \`is_mammal\` (a boolean) as parameters and initializes the respective instance attributes.
              2. A \`make_animal_sound\` method that is marked as an abstract method using \`raise NotImplementedError\`. This means that any concrete subclass of \`Animal\` must implement this method.

              To create an instance of a specific animal, you would create a subclass of \`Animal\` and implement the \`make_animal_sound\` method. For example:

              \`\`\`python
              class Dog(Animal):
                  def make_animal_sound(self) -> str:
                      return "Woof!"

              dog = Dog("Buddy", True)
              print(dog.name)  # Output: Buddy
              print(dog.is_mammal)  # Output: True
              print(dog.make_animal_sound())  # Output: Woof!
              \`\`\`

              Note that Python doesn't have a strict type system like TypeScript, so you can omit the type annotations if you prefer."
            `,
            explainPollyError
        )
    }, 30_000)

    it('commands/custom, chat command, no context', async () => {
        await client.openFile(animalUri)
        const result = (await client.request('commands/custom', {
            key: '/none',
        })) as CustomChatCommandResult
        expect(result.type).toBe('chat')
        const lastMessage = await client.firstNonEmptyTranscript(result.chatResult as string)
        expect(trimEndOfLine(lastMessage.messages.at(-1)?.text ?? '')).toMatchInlineSnapshot(
            `"no"`,
            explainPollyError
        )
    }, 30_000)

    // The context files are presented in an order in the CI that is different
    // than the order shown in recordings when on Windows, causing it to fail.
    it('commands/custom, chat command, current directory context', async () => {
        await client.openFile(animalUri)
        const result = (await client.request('commands/custom', {
            key: '/countDirFiles',
        })) as CustomChatCommandResult
        expect(result.type).toBe('chat')
        const lastMessage = await client.firstNonEmptyTranscript(result.chatResult as string)
        const reply = trimEndOfLine(lastMessage.messages.at(-1)?.text ?? '')
        expect(reply).not.includes('.cody/ignore') // file that's not located in the src/directory
        expect(reply).toMatchInlineSnapshot(`"9"`, explainPollyError)
    }, 30_000)

    it('commands/custom, edit command, insert mode', async () => {
        await client.openFile(sumUri, { removeCursor: false })
        const result = (await client.request('commands/custom', {
            key: '/hello',
        })) as CustomEditCommandResult
        expect(result.type).toBe('edit')
        await client.taskHasReachedAppliedPhase(result.editResult)

        const originalDocument = client.workspace.getDocument(sumUri)!
        expect(trimEndOfLine(originalDocument.getText())).toMatchInlineSnapshot(
            `
          "// hello
          export function sum(a: number, b: number): number {
              /* CURSOR */
          }"
        `,
            explainPollyError
        )
    }, 30_000)

    it('commands/custom, edit command, edit mode', async () => {
        await client.openFile(animalUri)

        const result = (await client.request('commands/custom', {
            key: '/newField',
        })) as CustomEditCommandResult
        expect(result.type).toBe('edit')
        await client.taskHasReachedAppliedPhase(result.editResult)

        const originalDocument = client.workspace.getDocument(animalUri)!
        expect(trimEndOfLine(originalDocument.getText())).toMatchInlineSnapshot(`
          "export interface Animal {
              name: string
              makeAnimalSound(): string
              isMammal: boolean
              logName(): void {
                  console.log(this.name);
              }
          }"
        `)
    }, 30_000)
})
