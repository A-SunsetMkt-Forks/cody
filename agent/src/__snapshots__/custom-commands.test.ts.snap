// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Custom Commands > commands/custom, chat command, open tabs context 1`] = `
"Based on the given context, here is a high-level overview and the names of the files you have shared so far, formatted as a bulleted list with only the raw filenames:

High-level Overview:
The provided context consists of three code snippets, each exporting a function named \`example\` that returns the string 'example'. These code snippets appear to be from separate files within a codebase.

File Names:
- example1.ts
- example2.ts
- example3.ts"
`;

exports[`Custom Commands > commands/custom, edit command, edit mode 1`] = `
"export interface Animal {
    name: string
    makeAnimalSound(): string
    isMammal: boolean
    logName(): void {
        console.log(\`Animal name: \${this.name}\`)
    }
}

"
`;

exports[`Custom Commands > commands/custom, edit command, insert mode 1`] = `
"// hello
export function sum(a: number, b: number): number {
    /* CURSOR */
}
"
`;
